cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractDistributedProgramRunner.java

```diff
@@ -231,15 +231,23 @@ public TwillController launch(TwillApplication twillApplication, Iterable<String
 
   private ProgramOptions addArtifactPluginFiles(ProgramOptions options, Map<String, LocalizeResource> localizeResources,
                                                 File tempDir) throws IOException {
-    File localDir = new File(options.getArguments().getOption(ProgramOptionConstants.PLUGIN_DIR));
-    File archiveFile = new File(tempDir, Constants.Plugin.DIRECTORY);
-    BundleJarUtil.packDirFiles(localDir, Locations.toLocation(archiveFile), tempDir);
-    localizeResources.put(Constants.Plugin.DIRECTORY, new LocalizeResource(archiveFile, true));
-    LOG.debug("Localizing Resource {} here {}", localDir, archiveFile);
-
-    Map<String, String> systemArgs = Maps.newHashMap(options.getArguments().asMap());
-    systemArgs.put(ProgramOptionConstants.PLUGIN_DIR, Constants.Plugin.DIRECTORY);
-    return new SimpleProgramOptions(options.getName(), new BasicArguments(systemArgs),
+    Arguments systemArgs = options.getArguments();
+    if (!systemArgs.hasOption(ProgramOptionConstants.PLUGIN_DIR)) {
+      return options;
+    }
+
+    File localDir = new File(systemArgs.getOption(ProgramOptionConstants.PLUGIN_DIR));
+    File archiveFile = new File(tempDir, "artifacts.jar");
+    BundleJarUtil.createJar(localDir, archiveFile);
+
+    // Localize plugins to two files, one expanded into a directory, one not.
+    localizeResources.put("artifacts", new LocalizeResource(archiveFile, true));
```

Comment:
The keys constants are needed as the remote program runner depends solely on the values (by looking up through the constant keys) to determine the location of the artifact directory and artifact jar.